### DISCUSSION TITLE
Use of python libraries

### DISCUSSION THREAD

**Discussion by IdenRosenthal** [codewars/codewars.com#2781] [Polls]
I have worked a kata that asked me to find all unique combinations of n elements of an array. There is a python library that takes care of this onerous task for you. (itertools.combinations). Is it "cheating" to use such a helper. Because in the real world we would have this at our disposal. But is it more to the point to do things "the hard way"?

**Comment by R4qun3** [comment]
I can't see how it can be  seen as cheating. Libraries are part of the language, therefore a valid part of a solution. Or am I wrong?

**Reply by Ebatdev34** [reply]
you're right but also knowing the logic by not using libraries make you understand more

**Reply by R4qun3** [reply]
Very good and valid point. Especially if your goal is to gain more granular understanding of the language. But if your goal for using these challenges is to become more proficient and, I guess 'economic', then I think using libraries should be fine. But that's just my perspective. (And I'm not that smart, lol)

**Comment by s1ddiq** [comment]
yuh

**Reply by xlebovoz** [reply]
айайайай как нехорошо

**Comment by mo-ibik09** [comment]
probably yes.

**Reply by isaiah777-sys** [reply]
It actually depends on what you're using it for
Plus I think it's better to use some. cos they make your work easier

**Comment by TopeMe** [comment]
It depends on your goal, if you want to solve it faster then use a library. But if you want to learn more and practice your logic tackling the task than you can go sticks and stones with it.

**Reply by R4qun3** [reply]
I just replied to someone in my original comment saying this exact thing, then saw your comment. I was doubting my opinion, but I feel better about it now, haha.

**Comment by manevdave** [comment]
Python libraries are vital tools that enhance development efficiency by providing pre-written code for common tasks. They facilitate data analysis (Pandas, NumPy), visualization (Matplotlib, Seaborn), and machine learning (Scikit-learn, TensorFlow). For web development, Flask and Django simplify application creation, while Requests and Beautiful Soup assist with automation and web scraping. Libraries like Pillow and OpenCV handle image processing, and SciPy supports scientific computing. Additionally, NLTK and spaCy are used for natural language processing, and testing frameworks like PyTest ensure code quality. Overall, these libraries boost productivity and functionality in various projects.

**Comment by ClipClopBoom** [comment]
Spirt vs letter I guess; in programming generally if you can solve a problem with a library then it makes sense. There are times when you need to roll your own (for example our production code has a very restricted list of approved libraries for security reasons), that and it’s a good learning exercise; too many learning resources take the lazy route and just have you calling a library function for everything and you never really look into how it works under the hood.

**Comment by Heoralight** [comment]
Sometimes it is mainly used to improve efficiency, but the attitude is subjective.
Using tools and libraries designed for specific tasks can save time and reduce the likelihood of errors. It's about finding a balance between understanding fundamental concepts and utilizing available resources effectively.

**Comment by A2nyc** [comment]
Ever since I heard of some libraries having malicious code, I think it's better to build the function yourself and only use libraries if you know what is in them.

**Comment by ravshanovdev** [comment]
hello hello everyone

**Comment by xlebovoz** [comment]
I don't understand how you can work without libraries, libraries greatly simplify work and, moreover, they were created to solve various complex problems

**Reply by krishnadalal29** [reply]
Libraries simplify and speed up development by solving complex problems with reusable code. However, developers may avoid them due to learning, performance, security, or resource constraints.

**Comment by aryuemaan** [comment]
**No, it's not cheating in the real world, and often not in a coding challenge either.**

In professional programming, the goal is typically to write correct, efficient, readable, and maintainable code as quickly as possible. If a standard library function like this so don't worry about it !

**Comment by iris-is-me** [comment]
I don't see a problem if the kata did not say you cannot use a module
